### PR TITLE
Create CFGVisualizeOptions class for handling command-line arguments

### PR DESCRIPTION
**Pull request by farzanmirshekari**
The CFGVisualizeArgs class has been created to abstract away the functionality pertaining to parsing command line arguments for CFG visualization.

Things to consider:
- Is CFGVisualizeArgs an appropriate name?
- Do we want printUsage() to be able to be set by the class using CFGVisualizeArgs? i.e. each user class can have its own usage instructions that are passed to the CFGVisualizeArgs instance?
- How should printError be handled? There is need for it in both CFGVisualizeArgs and CFGVisualizeLauncher, and potentially other classes as well.

Once the changes are reviewed, I can bring in CFGVisualizeArgs to all other relevant classes as well.

### CODE REVIEW

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeArgs.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+import java.io.File;
+
+public class CFGVisualizeArgs {
+
+    private static final String DEFAULT_METHOD = "test";
+    private static final String DEFAULT_CLASS = "Test";
+    private static final String DEFAULT_OUTPUT_DIR = ".";
+
+    public String input;
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeArgs.java:13]: All of these fields can be private, right?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeArgs.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
+package org.checkerframework.dataflow.cfg.visualize;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.File;
+
+public class CFGVisualizeArgs {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeArgs.java:7]: How about `CFGVisualizeOptions`?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeArgs.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        this.clas = clas;
+        this.pdf = pdf;
+        this.verbose = verbose;
+        this.string = string;
+    }
+
+    public static CFGVisualizeArgs parseArgs(String[] args) {
+        if (args.length == 0) {
+            printUsage();
+            System.exit(1);
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeArgs.java:41]: Mention that methods call `System.exit` in the class documentation.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeArgs.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        if (error) {
+            System.exit(1);
+        }
+
+        return new CFGVisualizeArgs(input, output, method, clas, pdf, verbose, string);
+    }
+
+    /** Print usage information. */
+    private static void printUsage() {
+        System.out.println(
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeArgs.java:110]: Maybe also mention that information is sent to `System.out/error`.

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks, this looks great! The use in LVPlayground makes it much nicer.
I have a few small suggestions. Please also add the missing javadoc to make the `misc` tasks pass.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java [hunk 5]
```diff
@@ -6,29 +6,32 @@
...
      */
     public static void main(String[] args) {
 
-        /* Configuration: change as appropriate */
-        String inputFile =
-                "dataflow/manual/examples/BusyExprSimple.java"; // input file name and path
-        String outputDir = "."; // output directory
-        String method = "test"; // name of the method to analyze
-        String clazz = "Test"; // name of the class to consider
+        /** Parse the arguments. */
```
[the remark this answers is not in the document]
Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:21]: Javadoc-style comments shouldn't appear in method bodies. Having a comment at all is questionable, given that the next line is pretty self-explanatory... I'll keep the comments for now, but feel free to remove them.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ConstantPropagationPlayground.java [hunk 6]
```diff
@@ -18,21 +19,24 @@ private ConstantPropagationPlayground() {
...
+     * @param args command-line arguments
      */
     public static void main(String[] args) {
 
-        /* Configuration: change as appropriate */
-        String inputFile = "Test.java"; // input file name and path
-        String outputDir = "cfg"; // output directory
-        String method = "test"; // name of the method to analyze
-        String clazz = "Test"; // name of the class to consider
+        /** Parse the arguments. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ConstantPropagationPlayground.java:26]:
```suggestion
        // Parse the arguments.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/LiveVariablePlayground.java [hunk 7]
```diff
@@ -13,21 +14,24 @@ public class LiveVariablePlayground {
...
+     * @param args command-line arguments
      */
     public static void main(String[] args) {
 
-        /* Configuration: change as appropriate */
-        String inputFile = "Test.java"; // input file name and path
-        String outputDir = "cfg"; // output directory
-        String method = "test"; // name of the method to analyze
-        String clazz = "Test"; // name of the class to consider
+        /** Parse the arguments. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/LiveVariablePlayground.java:21]:
```suggestion
        // Parse the arguments.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ReachingDefinitionPlayground.java [hunk 8]
```diff
@@ -12,22 +13,24 @@ public class ReachingDefinitionPlayground {
...
      */
     public static void main(String[] args) {
 
-        /* Configuration: change as appropriate */
-        String inputFile =
-                "./dataflow/manual/examples/ReachSimple.java"; // input file name and path
-        String outputDir = "."; // output directory
-        String method = "test"; // name of the method to analyze
-        String clazz = "Test"; // name of the class to consider
+        /** Parse the arguments. */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ReachingDefinitionPlayground.java:20]:
```suggestion
        // Parse the arguments.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+ * org.checkerframework.dataflow.cfg.visualize.CFGVisualizeLauncher} for an example.
+ */
+public class CFGVisualizeOptions {
+
+    /** default method name */
+    private static final String DEFAULT_METHOD = "test";
+    /** default class name */
+    private static final String DEFAULT_CLASS = "Test";
+    /** default output directory */
+    private static final String DEFAULT_OUTPUT_DIR = ".";
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java:22]:
```suggestion
    /** Default method name. */
    private static final String DEFAULT_METHOD = "test";
    /** Default class name. */
    private static final String DEFAULT_CLASS = "Test";
    /** Default output directory. */
    private static final String DEFAULT_OUTPUT_DIR = ".";
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    /** The method name. */
+    private String method;
+    /** The class name. */
+    private String clas;
+    /** True if the PDF should be generated. */
+    private boolean pdf;
+    /** True if the verbose output should be generated. */
+    private boolean verbose;
+    /** True if the string representation should be generated. */
+    private boolean string;
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java:37]:
```suggestion
    /** The input file. */
    private final String input;
    /** The output directory. */
    private final String output;
    /** The method name. */
    private final String method;
    /** The class name. */
    private final String clas;
    /** True if the PDF should be generated. */
    private final boolean pdf;
    /** True if the verbose output should be generated. */
    private final boolean verbose;
    /** True if the string representation should be generated. */
    private final boolean string;
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+
+        return new CFGVisualizeOptions(input, output, method, clas, pdf, verbose, string);
+    }
+
+    /**
+     * Getter for the input file.
+     *
+     * @return the input file
+     */
+    public String getInput() {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java:154]:
```suggestion
    public String getInputFile() {
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        return input;
+    }
+
+    /**
+     * Getter for the output directory.
+     *
+     * @return the output directory
+     */
+    public String getOutput() {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java:163]:
```suggestion
    public String getOutputDirectory() {
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+        return output;
+    }
+
+    /**
+     * Getter for the method name.
+     *
+     * @return the method name
+     */
+    public String getMethod() {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java:172]:
```suggestion
    public String getMethodName() {
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+        return method;
+    }
+
+    /**
+     * Getter for the class name.
+     *
+     * @return the class name
+     */
+    public String getClas() {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeOptions.java:181]:
```suggestion
    public String getClassName() {
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java [hunk 15]
```diff
@@ -6,29 +6,32 @@
...
         BusyExprTransfer transfer = new BusyExprTransfer();
         BackwardAnalysis<UnusedAbstractValue, BusyExprStore, BusyExprTransfer> backwardAnalysis =
                 new BackwardAnalysisImpl<>(transfer);
         CFGVisualizeLauncher.generateDOTofCFG(
-                inputFile, outputDir, method, clazz, true, true, backwardAnalysis);
+                config.getInput(),
+                config.getOutput(),
+                config.getMethod(),
+                config.getClas(),
+                true,
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:33]: Throughout, shouldn't these `true, true` be replaced with `config.isPDF(), config.isVerbose()`? Check the order of the two booleans.

Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:33]: Also, `config.isString()` isn't used here. How about we add a new method to `CFGVisualizeLauncher` that takes `transfer`, `analysis`, and `config` and then does the right thing?

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks for all these improvements! I have a bunch of small edits and only two small suggestions for you to address.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java (hunk 15 above)
Comment (farzanmirshekari) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:33]: Understood, I will make the changes! My mistake :)

Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:33]: You can simply commit the suggested edits directly on GitHub - add them to a batch and then commit them. When you edit more locally, remember to pull in from remote first.

Comment (farzanmirshekari) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:33]: Is `trasnfer` a parameter relevant to the methods inside `CFGVisualizeLauncher`? I was thinking of having a method called `performAnalysis` that takes in a `CFGVisualizeOptions` config object as well as an `analysis` object. In the playgrounds, the `transfer` object is used to generate the `analysis`. My current declaration for the function is:
```
public static void performAnalysis(
	    CFGVisualizeOptions config,
	    @Nullable Analysis<V, S, T> analysis
)
```

Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:33]: Yes, config and analysis should be enough.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java [hunk 16]
```diff
@@ -6,29 +6,25 @@
 import org.checkerframework.dataflow.busyexpr.BusyExprStore;
 import org.checkerframework.dataflow.busyexpr.BusyExprTransfer;
 import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeLauncher;
+import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeOptions;
 
 /** The playground for busy expression analysis */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:11]:
```suggestion
/** The playground for busy expression analysis.
 * As an example, try {@code dataflow/manual/examples/BusyExprSimple.java}.
 */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java [hunk 17]
```diff
@@ -6,29 +6,25 @@
...
 public class BusyExpressionPlayground {
 
     /**
      * Run busy expression analysis playground on a test file and print the CFG graph
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:15]:
```suggestion
     * Run busy expression analysis on a file.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ConstantPropagationPlayground.java [hunk 18]
```diff
@@ -18,21 +19,17 @@ private ConstantPropagationPlayground() {
     /**
      * Run constant propagation for a specific file and create a PDF of the CFG.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ConstantPropagationPlayground.java:20]:
```suggestion
     * Run constant propagation analysis on a file.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/LiveVariablePlayground.java [hunk 19]
```diff
@@ -13,21 +14,17 @@ public class LiveVariablePlayground {
     /**
      * Run live variable analysis for a specific file and create a PDF of the CFG in the end.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/LiveVariablePlayground.java:15]:
```suggestion
     * Run live variable analysis on a file.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeLauncher.java [hunk 20]
```diff
@@ -47,80 +46,78 @@ private CFGVisualizeLauncher() {
...
-                    break;
-                case "--string":
-                    string = true;
-                    break;
-                default:
-                    printError("Unknown command line argument: " + args[i]);
-                    error = true;
-                    break;
+    /**
+     * Generate the DOT representation of the CFG for a method, with a given analysis.
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeLauncher.java:58]:
```suggestion
     * Generate a visualization of the CFG of a method, with an optional analysis.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeLauncher.java [hunk 21]
```diff
@@ -47,80 +46,78 @@ private CFGVisualizeLauncher() {
...
+                                config.getMethodName(),
+                                config.getClassName(),
+                                config.isVerbose(),
+                                analysis);
+                if (res != null) {
+                    String stringGraph = (String) res.get("stringGraph");
+                    if (stringGraph == null) {
+                        System.err.println(
+                                "Unexpected output from generating string control flow graph, shouldn't be"
+                                        + " null.");
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeLauncher.java:111]:
```suggestion
                                        + " null. Result map: " + res);
```

**Review by wmdietl — APPROVED** [review]
Thanks! The CI failures are expected and don't block merging.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java [hunk 22]
```diff
@@ -6,29 +6,27 @@
 import org.checkerframework.dataflow.busyexpr.BusyExprStore;
 import org.checkerframework.dataflow.busyexpr.BusyExprTransfer;
 import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeLauncher;
+import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeOptions;
 
-/** The playground for busy expression analysis */
+/** The playground for busy expression analysis.
+ * As an example, try {@code dataflow/manual/examples/BusyExprSimple.java}.
+ */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:13]:
```suggestion
/**
 * The playground for busy expression analysis. As an example, try {@code
 * dataflow/manual/examples/BusyExprSimple.java}.
 */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ReachingDefinitionPlayground.java [hunk 23]
```diff
@@ -4,30 +4,28 @@
...
 import org.checkerframework.dataflow.analysis.UnusedAbstractValue;
 import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeLauncher;
+import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeOptions;
 import org.checkerframework.dataflow.reachingdef.ReachingDefinitionStore;
 import org.checkerframework.dataflow.reachingdef.ReachingDefinitionTransfer;
 
-/** The playground of reaching definition analysis. */
+/** The playground for reaching definition analysis.
+ * As an example, try {@code dataflow/manual/examples/ReachSimple.java}.
+ */
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ReachingDefinitionPlayground.java:13]:
```suggestion
/**
 * The playground for reaching definition analysis. As an example, try {@code
 * dataflow/manual/examples/ReachSimple.java}.
 */
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeLauncher.java [hunk 24]
```diff
@@ -47,80 +46,78 @@ private CFGVisualizeLauncher() {
...
+                                config.getMethodName(),
+                                config.getClassName(),
+                                config.isVerbose(),
+                                analysis);
+                if (res != null) {
+                    String stringGraph = (String) res.get("stringGraph");
+                    if (stringGraph == null) {
+                        System.err.println(
+                                "Unexpected output from generating string control flow graph, shouldn't be"
+                                        + " null. Result map: " + res);
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/visualize/CFGVisualizeLauncher.java:111]:
```suggestion
                                        + " null. Result map: "
                                        + res);
```

**Review by wmdietl — COMMENTED** [review]
Format fixes...

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java [hunk 25]
```diff
@@ -6,29 +6,28 @@
...
 import org.checkerframework.dataflow.busyexpr.BusyExprTransfer;
 import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeLauncher;
+import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeOptions;
 
-/** The playground for busy expression analysis */
+/**
+ * The playground for busy expression analysis. As an example, try {@code
+ * dataflow/manual/examples/BusyExprSimple.java}.
+ */
 public class BusyExpressionPlayground {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/BusyExpressionPlayground.java:15]:
```suggestion
public class BusyExpressionPlayground {

    /** Class cannot be instantiated. */
    private BusyExpressionPlayground() {
        throw new AssertionError("Class BusyExpressionPlayground cannot be instantiated.");
    }
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/LiveVariablePlayground.java [hunk 26]
```diff
@@ -4,30 +4,27 @@
 import org.checkerframework.dataflow.analysis.BackwardAnalysisImpl;
 import org.checkerframework.dataflow.analysis.UnusedAbstractValue;
 import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeLauncher;
+import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeOptions;
 import org.checkerframework.dataflow.livevariable.LiveVarStore;
 import org.checkerframework.dataflow.livevariable.LiveVarTransfer;
 
 /** The playground of live variable analysis. */
 public class LiveVariablePlayground {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/LiveVariablePlayground.java:12]:
```suggestion
public class LiveVariablePlayground {

    /** Do not instantiate. */
    private LiveVariablePlayground() {
        throw new Error("do not instantiate");
    }
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ReachingDefinitionPlayground.java [hunk 27]
```diff
@@ -4,30 +4,29 @@
...
+import org.checkerframework.dataflow.cfg.visualize.CFGVisualizeOptions;
 import org.checkerframework.dataflow.reachingdef.ReachingDefinitionStore;
 import org.checkerframework.dataflow.reachingdef.ReachingDefinitionTransfer;
 
-/** The playground of reaching definition analysis. */
+/**
+ * The playground for reaching definition analysis. As an example, try {@code
+ * dataflow/manual/examples/ReachSimple.java}.
+ */
 public class ReachingDefinitionPlayground {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/playground/ReachingDefinitionPlayground.java:15]:
```suggestion
public class ReachingDefinitionPlayground {

    /** Class cannot be instantiated. */
    private ReachingDefinitionPlayground() {
        throw new AssertionError("Class ReachingDefinitionPlayground cannot be instantiated.");
    }

```

**Review by wmdietl — COMMENTED** [review]
Update default constructors

**Pull request merged**